Spec abi chapter (#1545)
rust-lang/reference: src/abi.md

> [!NOTE]
> This can include calls to functions defined outside of Rust, or built using a different Rust compiler version.
> Additional guarantees will apply in this case for "FFI Safe" types, which match up with the platform C ABI in well-defined ways.
> These are not fully documented here currently.

> [!WARNING]
> Two types that are ABI Compatible may not allow the same set of values (e.g. [`*const T`] and [`core::ptr::NonNull<T>`]).
> If an invalid value is passed as a parameter or returned from a function, the result is immediate undefined behaviour, even if the parameter or return value is never used.

> > For example, passing a null pointer to a function that accepts a `NonNull<i32>` parameter via `fn(*const i32)` function pointer caused undefined behaviour.

Reviewer: ```suggestion
> For example, passing a null pointer to a function that accepts a `NonNull<i32>` parameter via `fn(*const i32)` function pointer causes undefined behavior.
```

# Application Binary Interface (ABI)

r[abi]

## ABI Compatibility

> r[abi.compatibility]

Reviewer 2: Can you add a few sentences explaining what *abi compatibility* is? Such as the consequence of what it means for types to be abi compatible? In general, I felt a little uncertain what I was reading, since it immediately jumps into a list of rules without giving context or explaining what this section is about.

r[abi.compatibilty.type]
Two types, `T` and `U`, can be *abi compatible*.

r[abi.compatibility.equivalence]

> Two types `T` and `U` are *abi compatible* if:

Reviewer 3: My first reaction upon seeing these three bullets was to say "yes this yields an equivalence relation, but not an interesting one". Skimming down I was reminded that there were going to be other base cases in the inductive definition, which is what makes it interesting.

So: Is there a (concise) way to emphasize that this is just an "if" and not an "if and only if" ?

* They are the same type,
* `U` is *abi compatible* with `T`, or

> * There exists a type `V`, such that `T` is *abi compatible* with `V` an `V` is *abi compatuble* with `U`,

Reviewer 4: ```suggestion
* There exists a type `V`, such that `T` is *abi compatible* with `V` an `V` is *abi compatible* with `U`,
```

> [!NOTE]

> > These properties ensure that *abi compatibility* is an equivalence relation.

Author: Probably a good idea. I wouldn't want to put it on the bullets though, maybe add it to the note.

> These properties are respectively called "reflexivity", "symmetry", and "transitivity". They ensure that *abi compatibility* is an equivalence relation.

r[abi.compatibility.integer]

> Two [integer types] are *abi compatible* if they have the same size and the same signednes

Reviewer 4: ```suggestion
Two [integer types] are *abi compatible* if they have the same size and the same signedness
```

> [!NOTE]
> In particular, `usize` is *abi compatible* with `uN`, and `isize` is *abi compatible* with `iN` where `N` is the target_pointer_width.
> Two integer types with different signedness, such as `u8` and `i8` are not *abi compatible*.

```rust
#[cfg(target_pointer_width="32")]
fn foo(x: u32) -> u32{
    x
}
#[cfg(target_pointer_width="64")]
fn foo(x: u64) -> u64{
    x
}

fn main(){
    let f: fn(usize)->usize = unsafe{core::mem::transmute(foo as fn(_)->_)};
    let x = 0usize;
    let y = f(x);
    assert_eq!(x,y);
}
```

r[abi.compatibility.char]
The type [`char`] is *abi compatible* with the type [`u32`][integer types].

```rust
fn foo(x: char) -> u32{
    x as u32
}

fn main(){
    let f: fn(u32)->char = unsafe{core::mem::transmute(foo as fn(_)->_)};
    let x = b'A' as u32; // ascii character indecies are the same as Unicode character indecies
    let y = f(x);
    assert_eq!(y, 'A');
}
```

r[abi.compatibility.pointer]

> Two [pointer types], `*mut T` and `*const U`, are *abi compatible* if the *metadata type*s of `T` and `U` are the same type.

Author: Will be defined as part of layout.  The term is used in the docs.rs page.

> [!NOTE]
> [`core::marker::Sized`] types have a *metadata type* of `()`.

> [!NOTE]
> With transitivity, this applies regardless of the mutability of either pointer type

```rust
unsafe fn foo(x: *mut i32){
   unsafe{x.write(5);}
}

fn main(){
    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function

>     let mut val = 0;

Reviewer 3: this example is relying on `i32` being the fallback integer type when there is no other alternative being imposed, right?

(E.g., if one had written `5_i8` down below, then that ends up affecting the type inferred for `val`, and yields UB overall since now the write will be out-of-bounds, at least according to miri.)

I am wondering whether it would be better, for purposes of this example, to explicitly assign the `i32` type via `let val: i32 = 0;` and then you avoid discussion of how integer type fallback is handled in this part of the spec.

Author: Probably, yeah. I guess my brain was just on autopilot writing that test.

```rust
fn foo(x: &mut i32){
   *x = 5;
}

fn main(){
    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as fn(_))}; // Type Erase the function
    let mut val = 0;
    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
    unsafe{f(ptr);}
    assert_eq!(val, 5);
}
```

r[abi.compatibility.core]
The types [`core::mem::MaybeUninit<T>`], [`core::cell::UnsafeCell<T>`], and [`core::num::NonZero<T>`], are *abi compatible* with `T`.

r[abi.compatibility.transparent]
A [`struct`] declared with the `transparent` representation is *abi compatible* with its field that does not have size 0 and alignment 1, if such a field exists.

r[abi.compatibilty.zst]
Two types, `T` and `U`, are *abi compatible* if both have size 0 and alignment 1.

r[abi.compatibility.option]
If `T` is a type listed in [layout.enum.option](https://doc.rust-lang.org/stable/core/option/index.html#representation), then given `S` is a type with size 0 and alignment 1, `T` is *abi compatible* with the types [`core::option::Option<T>`], [`core::result::Result<T,S>`], and [`core::result::Result<S,T>`].

r[abi.compatibility.fn-ptr]

> An [`fn`-ptr type] `T` is compatible with an [`fn`-ptr type] `U` if `T` and `U` have *abi compatible* tags.

Reviewer 3: is this supposed to say "*abi compatible*", or just "compatible" ?

r[abi.compatibility.extern-tag]
Two [abi tags][abi tag] are *abi compatible* if:
* They are the same string, or
* One tag is `"X"`, and the other is `"X-unwind"`

r[abi.compatibility.signature]
Two function signatures are compatible if:
* The [abi tags][abi tag] of both signatures are *abi compatible*,
* They have the same number of parameters, excluding C-varargs,
* Each parameter of both signatures, in order, are *abi compatible*, and
* Either both signatures have C-varargs, or neither signature does.

> [!NOTE]
> A signature is compatible with itself.

r[abi.compatibility.simd-abi]

> A type has *simd abi requirements* if:

Reviewer 2: Can you add something here that explains what it means to have *simd abi requirements*? It seems to immediately jump in to the rules for what has them, but doesn't seem to really explain what the consequence is or what it is used for. I think a few sentences here could help a lot.

Author: It is stated in the following note.

* It is a type declared with the standard-library repr-attrbute `simd`,

> * It is a aggregate type[^1], which has a type with *simd abi requirements* as a field.

Reviewer 2: Can you use labels instead of numbers for footnotes? It makes it a little easier to avoid collisions, and makes it easier to see semantically what a footnote is referring to.

> [!TARGET-SPECIFIC]
> On x86 and x86-64, the [*salient target features*][target_feature] of the `simd` types are:
> * [`__m128`], [`__m128i`], [`__m128f`], and [`__m128d`] (128-bit vector types): `sse`
> * [`__m256`], [`__m256i`], [`__m256f`], and [`__m256d`] (256-bit vector types): `avx`
> * [`__m512`], [`__m512i`], [`__m512f`], and [`__m512d`] (512-bit vector types): `avx512f` and `avx512vl`

r[abi.compatibility.call]
A call to a function `f` via a function item or function pointer with a given signature `S` is valid if and only if the signature of the definition `f` is *compatible* with the signature `S`, and:
* The ABI tag of the signature is `extern "Rust"`, or
* If any parameter type, the return type, or the type of any argument passed via C-varargs has *simd abi requirements*, each [*salient target features*][target_feature] of that type is either set at both the definition site of the function, and at the call site, or is set at neither site.

> The behaviour a call that is not valid is undefined.

Reviewer 2: Can you say where you got these rules from?

Author: From: https://doc.rust-lang.org/core/primitive.fn.html#abi-compatibility

> [!NOTE]
> When parameter/return types do not exactly match, they are converted as though by calling [`core::mem::transmute`]. The representation and validity requirements of the type in the definition/return site still apply, for example, passing `0` to a function pointer `fn(u32)` that points to a function declared as `fn foo(x: NonZeroU32)` is undefined behaviour.

> [!NOTE]

> > the ABI tag `extern "Rust"` is the default when the `extern` keyword is not used (either to declare the function within an [`extern` block], or as a [function qualifier][extern functions]). Thus it is safe to call most functions that use simd types.

Reviewer 2: ```suggestion
> The ABI tag `extern "Rust"` is the default when the `extern` keyword is not used (either to declare the function within an [`extern` block], or as a [function qualifier][extern functions]). Thus it is safe to call most functions that use simd types.
```

[^1]: The aggregate types, for the purposes of this clause, are [`struct`] types, [`enum`] types, [`union`] types, and [array] types.

[`__m128`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m128.html
[`__m128i`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m128i.html
[`__m128f`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m128f.html
[`__m128d`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m128d.html
[`__m256`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m256.html
[`__m256i`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m256i.html
[`__m256f`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m256f.html
[`__m256d`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m256d.html
[`__m512`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m512.html
[`__m512i`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m512i.html
[`__m512f`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m512f.html
[`__m512d`]: https://doc.rust-lang.org/stable/core/arch/x86_64/struct.__m512d.html

## The `used` attribute

r[abi.used]

> ```abnf
> MetaItemUsed := "used"
> ```

Reviewer 2: Can you stick with the abnf style used in the rest of the reference? I realize it's not the easiest thing to write, but I would prefer to avoid having various different styles.

```suggestion
> **<sup>Attribute Syntax</sup>**\
> _MetaItemUsed_ :\
> &nbsp;&nbsp; `used`
```

> 

Reviewer 2: Can you add a sentence or two here introducing the attribute and briefly explaining what it is for. Generally the template for attributes that we follow is:

The *`NAME` attribute* ...explain what it is for.

Having introductory material is really helpful to give context and immediately understand what the following rules are talking about. It can be helpful to know the purpose up-front, and let the following rules further refine the details.

# Application Binary Interface (ABI)

r[abi]

This section documents features that affect the ABI of the compiled output of
a crate.

See *[extern functions]* for information on specifying the ABI for exporting
functions. See *[`extern` block]s* for information on specifying the ABI for
linking external libraries.

## ABI compatibility

> r[abi.compatibility]

Reviewer 5: There's no paragraph after this marker... is this the normal syntax?

Author: Yes, it's an identifier for the whole section.

r[abi.compatibility.intro]

> Function calls pass parameters and return values between the caller and the callee function. This requires the caller and callee to agree on an ABI for those parameters and return values. This is typically only guaranteed when the same type is used in both the call site and the definition site of the callee, but certain other types may be *abi compatible*. This can appear when transmuting a [function pointer] or using an [`extern` block] to call a function. When the parameters or return types differ between the call site and def site, assuming they are *abi compatible*, the parameters or return types are transmuted to the type at the def site or call site respectively.

Reviewer 5: ```suggestion
Function calls pass parameters and return values between the caller and the callee function. This requires the caller and callee to agree on an ABI for those parameters and return values. This is typically only guaranteed when the same type is used in both the call site and the definition site of the callee. However, differences between caller and callee types can appear when transmuting a [function pointer] or using an [`extern` block] to call a function. This is permitted if the caller and callee types are *abi compatible* (otherwise, it is undefined behavior). If the types are abi compatible, the parameters are transmuted to the callee type as part of the call and the return value is transmuted to the caller type upon return.
```

> [!NOTE]

> > This can include calls to functions defined outside of rust, or built using a different Rust compiler version.

Reviewer 5: ```suggestion
> This can include calls to functions defined outside of Rust, or built using a different Rust compiler version.
```

> Additional guarantees will apply in this case for "FFI Safe" types, which match up with the platform C ABI in well-defined ways.

> > These are not fully documented here currently.

Reviewer 5: Shouldn't this say that additional *requirements* apply? The document right now only describes the rules for Rust-to-Rust calls, as far as I can see.

Author: No, because the note is talking about the fact that "FFI Safe" types have a defined mapping to the platform C abi. 
You can technically pass a non-"FFI Safe" type to an FFI call, there's just no guarantee you can match it on the other side.

Reviewer 5: Oh so this is about guarantees of the form "This Rust type has an ABI matching this C type". Would be good to clarify the comment here, I currently find it a bit confusing -- everything else here is about guarantees of the form "these two Rust types have the same ABI", after all.

> [!WARNING]
> Two types that are ABI Compatible may not allow the same set of values (e.g. [`*const T`] and [`core::ptr::NonNull<T>`]).
> If an invalid value is passed as a parameter or returned from a function, the result is immediate undefined behaviour, even if the parameter or return value is never used.
> For example, passing a null pointer to a function that accepts a `NonNull<i32>` parameter via `fn(*const i32)` function pointer caused undefined behaviour.

r[abi.compatibility.equivalence]
Two types `T` and `U` are *abi compatible* if:
* They are the same type,
* `U` is *abi compatible* with `T`, or

> * There exists a type `V`, such that `T` is *abi compatible* with `V` an `V` is *abi compatible* with `U`,

Reviewer 5: IMO this is better stated after all the other clauses, since really what we are doing is we are generating a relation by having some "seeds" and then we are upclosing it to become an equivalence relation. It is quite confusing to read this before we added any of the elements to the relation that make it interesting.

> [!NOTE]
> With transitivity, this applies regardless of the mutability of either pointer type

```rust
unsafe fn foo(x: *mut u32){
   unsafe{x.write(5);}
}

fn main(){
    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as unsafe fn(_))}; // Type Erase the function
    let mut val: u32 = 0;
    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
    unsafe{f(ptr);}
    assert_eq!(val, 5);
}
```

r[abi.compatibility.reference-box]
The types [`&T`], [`&mut T`], [`alloc::boxed::Box<T>`], and [`core::ptr::NonNull<T>`], are *abi compatible* with `*const T`

> [!NOTE]
> With transitivity, they are also *abi compatible* with each other, and with `*mut T`, as well as references/`Box` to different types that have the same *metadata type*.

```rust
fn foo(x: &mut i32){
   *x = 5;
}

fn main(){
    let f: unsafe fn(*mut ()) = unsafe{core::mem::transmute(foo as fn(_))}; // Type Erase the function
    let mut val = 0;
    let ptr = core::ptr::addr_of_mut!(val).cast::<()>(); // Get Opaque Userdata from somewhere
    unsafe{f(ptr);}
    assert_eq!(val, 5);
}
```

r[abi.compatibility.core]

> The types [`core::mem::MaybeUninit<T>`], [`core::cell::UnsafeCell<T>`], and [`core::num::NonZero<T>`], are *abi compatible* with `T`.

Reviewer 5: This is odd... why do we list these types and not others?

Author: These types are magic lang items. `MaybeUninit<T>` is a transparent union (which isn't otherwise a thing that can exist), `UnsafeCell<T>` doesn't fully qualify as normal `repr(transparent)` due to excluding niches (including guaranteed ones that would normally be inherited by `repr(transparent)`, and `NonZero<T>` has GDE.

r[abi.compatibility.transparent]
A [`struct`] declared with the `transparent` representation is *abi compatible* with its field that does not have size 0 and alignment 1, if such a field exists.

r[abi.compatibility.zst]
Two types, `T` and `U`, are *abi compatible* if both have size 0 and alignment 1.

r[abi.compatibility.discriminant]

> If `T` is an a type listed in [layout.repr.rust.option.elision], and `U` is the type of the *elision candidate field*, then `T` is layout compatible with `U`.

Reviewer 5: The referenced section here doesn't exist yet, does it?

Reviewer 5: Oh this is just a normal markdown link for now, I see.

Author: Getting rid of the hard dependency on #1654.
If #1654 lands first, then I'll remove the link. If this lands before #1654 I'll make #1645 remove the link.

> [!NOTE]
> `Option<U>`, `Result<U,Z>`, or `Result<Z,U>` are such types, when `U` are *elision candidate type*s, and `Z` is a 1-ZST type.
>
> Due to transitivity, two such types are *abi compatible* with each other if their *elision candidate field*s are *abi comaptible*

r[abi.compatibility.fn-ptr]

> An [function pointer] type `T` is *abi compatible* with an [function pointer] type `U` if `T` and `U` have *abi compatible* tags.

Reviewer 5: ```suggestion
A [function pointer] type `T` is *abi compatible* with an [function pointer] type `U` if `T` and `U` have *abi compatible* tags.
```

r[abi.compatibility.extern-tag]
Two [abi tags][abi tag] are *abi compatible* if:
* They are the same string, or
* One tag is `"X"`, and the other is `"X-unwind"`

> [!NOTE]
> e.g. `extern "C"` and `extern "C-unwind"` are compatible with each other.

r[abi.compatibility.signature]
Two function signatures are compatible if:
* The [abi tags][abi tag] of both signatures are *abi compatible*,
* They have the same number of parameters, excluding C-varargs,
* Each parameter of both signatures, in order, are *abi compatible*, and
* Either both signatures have C-varargs, or neither signature does.

> [!NOTE]
> A signature is compatible with itself.

> r[abi.compatibility.simd-abi]

Reviewer 5: Given that we consider this a bug and are very close to fixing the bug to at least emit warnings, I don't think we should list this here.

Reviewer 5: This clause no longer exists in the library docs, so it should be removed here as well.

Author: Yep, was waiting for the stdlib to fix to land. Will fix this tomorrow.

r[abi.compatibility.simd-target-feature-x86]
> [!TARGET-SPECIFIC]
> On x86 and x86-64, the [*salient target features*][target_feature] of the `simd` types are:
> * [`__m128`], [`__m128i`], [`__m128f`], and [`__m128d`] (128-bit vector types): `sse`
> * [`__m256`], [`__m256i`], [`__m256f`], and [`__m256d`] (256-bit vector types): `avx`
> * [`__m512`], [`__m512i`], [`__m512f`], and [`__m512d`] (512-bit vector types): `avx512f` and `avx512vl`

r[abi.compatibility.call]
A call to a function `f` via a function item or function pointer with a given signature `S` is valid if and only if the signature of the definition `f` is *compatible* with the signature `S`, and:
* The ABI tag of the signature is `extern "Rust"`, or
* If any parameter type, the return type, or the type of any argument passed via C-varargs has *simd abi requirements*, each [*salient target features*][target_feature] of that type is either set at both the definition site of the function, and at the call site, or is set at neither site.

> The behavior of a call that is not valid is undefined.

Reviewer 5: Should this say that a valid call transmutes the arguments / return value?

0000000000000000 T foo::quux
```

## Symbol naming

r[abi.symbol-name]

```abnf
MetaItemNoMangle := "no_mangle"
MetaItemExportName := "export_name" "=" ([STRING_LITERAL] | [RAW_STRING_LITERAL])
```

r[abi.symbol-name.names]

> The *`no_mangle` attribute* and the *`export_name` attribute* shall only be applied to a `static` or `fn` item. The *`export_name` attribute* shall not be applied to an item declared within an [`extern` block].

Reviewer 2: Can't no_mangle be applied to all other items?

```rust
#[no_mangle]
extern "C" fn foo(x: i32) -> i32 {
    x + 1
}

#[export_name = "bar"]
extern "C" fn baz(x: i32) -> i32 {
    x + 2
}
```

```rust,compile_fail
extern "C" {
    #[export_name = "foo"]

>     fn __foo(x: i32) -> i32;

Reviewer 2: It can be helpful for error examples to highlight where the error is.

```suggestion
    fn __foo(x: i32) -> i32;  // error: not a free function, impl method, or static
```

pub static VAR1: u32 = 1;
```

r[abi.link_section.def]

> An item with the *`link_section` attribute* is placed in the specified section when linking. The section specified shall not violate the constraints on section names on the target, and shall not be invalid for the item type, no diagnostic is required.

Reviewer 2: Similar to our conversation before about ndr, I think it would be best to avoid this for now until we have a decision on it.

> [!NOTE]

> > A section name may be invalid if it violates the requirements for the item type, for example, an `fn` item must be placed in an executable section, and a mutable static item (`static mut` or one containing an `UnsafeCell`) must be placed in a writable section.

Reviewer 2: ```suggestion
> A section name may be invalid if it violates the requirements for the item type. For example, an `fn` item must be placed in an executable section, and a mutable static item (`static mut` or one containing an `UnsafeCell`) must be placed in a writable section.
```
